src/org/jitsi/impl/neomedia/rtp/translator/SimulcastController.java

```diff
@@ -0,0 +1,532 @@
+
+            onAccept(buf, off, len);
+
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public void update(int targetIdx)
+    {
+        if (this.targetIdx == targetIdx)
+        {
+            return;
+        }
+
+        if (logger.isDebugEnabled())
+        {
+            long ssrc = weakSource.get().getRTPEncodings()[0].getPrimarySSRC();
+            logger.debug("target_update ssrc=" + ssrc + ",target=" + targetIdx);
+        }
+
+        this.targetIdx = targetIdx;
+        if (targetIdx < 0)
+        {
+            // suspend the stream.
+            transformState = dropState;
+        }
+        else
+        {
+            // send FIR.
+            MediaStreamTrackImpl sourceTrack = weakSource.get();
+            if (sourceTrack == null)
+            {
+                return;
+            }
+
+            ((RTPTranslatorImpl) sourceTrack.getMediaStreamTrackReceiver()
+                .getStream().getRTPTranslator())
+                .getRtcpFeedbackMessageSender().sendFIR((int) targetSSRC);
+        }
+    }
+
+    /**
+     * Increments the {@link #filterState} as a result of accepting the packet
+     * specified in the arguments.
+     *
+     * @param buf the <tt>byte</tt> array that holds the RTP packet.
+     * @param off the offset in <tt>buffer</tt> at which the actual RTP data
+     * begins.
+     * @param len the number of <tt>byte</tt>s in <tt>buffer</tt> which
+     * constitute the actual RTP data.
+     */
+    private void onAccept(byte[] buf, int off, int len)
+    {
+        long ts = transformState
+            .rewriteTimestamp(RawPacket.getTimestamp(buf, off, len));
+
+        if (filterState.maxTs == -1
+            || TimeUtils.rtpDiff(ts, filterState.maxTs) > 0)
+        {
+            filterState.maxTs = ts;
+        }
+
+        int seqNum = transformState
+            .rewriteSeqNum(RawPacket.getSequenceNumber(buf, off, len));
+
+        if (filterState.maxSeqNum == -1
+            || RTPUtils.sequenceNumberDiff(seqNum, filterState.maxSeqNum) > 0)
+        {
+            filterState.maxSeqNum = seqNum;
+        }
+
+        filterState.transmittedBytes += len;
+        filterState.transmittedPackets++;
+    }
+
+    /**
+     * Transforms outgoing RTP packets.
+     */
+    private class RTPTransformer implements PacketTransformer
+    {
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void close()
+        {
+            // TODO update the receive counters
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public RawPacket[] reverseTransform(RawPacket[] pkts)
+        {
+            //  No need for that (yet?).
+            return pkts;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public RawPacket[] transform(RawPacket[] pkts)
```

Comment:
(And the temporal scalability controller will be used by the simulcast controller)